freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/downscale/action/FreeIpaDownscaleActions.java

```diff
@@ -0,0 +1,331 @@
+import com.sequenceiq.freeipa.flow.freeipa.downscale.failure.RevokeCertsResponseToDownscaleFailureEventConverter;
+import com.sequenceiq.freeipa.flow.freeipa.provision.event.clusterproxy.ClusterProxyUpdateRegistrationRequest;
+import com.sequenceiq.freeipa.flow.freeipa.provision.event.clusterproxy.ClusterProxyUpdateRegistrationSuccess;
+import com.sequenceiq.freeipa.flow.stack.AbstractStackAction;
+import com.sequenceiq.freeipa.flow.stack.StackEvent;
+import com.sequenceiq.freeipa.flow.stack.termination.action.TerminationService;
+import com.sequenceiq.freeipa.service.operation.OperationService;
+import com.sequenceiq.freeipa.service.resource.ResourceService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+import com.sequenceiq.freeipa.service.stack.StackUpdater;
+
+@Configuration
+public class FreeIpaDownscaleActions {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaDownscaleActions.class);
+
+    @Inject
+    private StackUpdater stackUpdater;
+
+    @Inject
+    private InstanceMetaDataToCloudInstanceConverter instanceConverter;
+
+    @Inject
+    private ResourceService resourceService;
+
+    @Inject
+    private ResourceToCloudResourceConverter resourceConverter;
+
+    @Inject
+    private TerminationService terminationService;
+
+    @Bean(name = "STARTING_DOWNSCALE_STATE")
+    public Action<?, ?> startingDownscaleAction() {
+        return new AbstractDownscaleAction<>(DownscaleEvent.class) {
+            @Override
+            protected void doExecute(FreeIpaDownscaleContext context, DownscaleEvent payload, Map<Object, Object> variables) {
+                Stack stack = context.getStack();
+                List<String> instanceIds = context.getInstanceIds();
+                LOGGER.info("Starting downscale of stack {} instances [{}]", stack.getId(), instanceIds);
+                stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.DOWNSCALE_IN_PROGRESS, "Starting downscale");
+                sendEvent(context, STARTING_DOWNSCALE_FINISHED_EVENT.selector(), new StackEvent(stack.getId()));
+            }
+        };
+    }
+
+    @Bean(name = "DOWNSCALE_CLUSTERPROXY_REGISTRATION_STATE")
+    public Action<?, ?> downscaleClusterProxyRegistrationAction() {
+        return new AbstractDownscaleAction<>(StackEvent.class) {
+            @Override
+            protected void doExecute(FreeIpaDownscaleContext context, StackEvent payload, Map<Object, Object> variables) throws Exception {
+                Stack stack = context.getStack();
+                stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.DOWNSCALE_IN_PROGRESS, "Updating cluster proxy registration.");
+                List<String> instanceIdsToRegister = stack.getNotDeletedInstanceMetaDataList().stream()
+                        .map(InstanceMetaData::getInstanceId)
+                        .filter(instanceId -> !context.getInstanceIds().contains(instanceId))
```

Comment:
like here, the `AbstractDownscaleAction` could have a method for getting the instanceIds from the `variables`, and not put it in the context, because now we have the same info available in 2 places